File: implementations/csharp/client/tests/Specus.Client.Tests/SharedAdversarialVectorTests.cs
using System.Diagnostics;
using Specus.Client.PeerMesh;

namespace Specus.Client.Tests;

/// <summary>
/// The shared adversarial corpus in <c>protocol/test-vectors/adversarial-inputs.json</c>.
///
/// <para>One implementation rejecting a hostile input does not mean the input is handled: an
/// attacker picks whichever node runs the most permissive implementation. Java, Go, .NET and
/// Android all run this same file, so a gap in one of them shows up as a failure rather than as a
/// difference nobody thought to look for.</para>
///
/// <para>The requirement per case is either a clean rejection or a value the caller can safely
/// use — never an unhandled exception, and never so slow that a hostile packet could stall a
/// receive loop.</para>
/// </summary>
public sealed class SharedAdversarialVectorTests
{
    [Fact]
    public void EverySharedCaseIsHandled()
    {
        var document = ProtocolVectorTestHelper.Read<AdversarialVectors>(
            "protocol/test-vectors/adversarial-inputs.json");

        Assert.NotEmpty(document.Cases);

        foreach (var testCase in document.Cases)
        {
            var payload = Convert.FromHexString(testCase.PayloadHex);

            var stopwatch = Stopwatch.StartNew();
            DecodeAllReachable(payload);
            stopwatch.Stop();

            Assert.True(stopwatch.ElapsedMilliseconds < 1_000,
                $"{testCase.Name} took {stopwatch.ElapsedMilliseconds}ms to decide; a hostile "
                + "input must not stall a receive loop");
        }
    }

    /// <summary>
    /// Feeds one payload to every decoder a datagram can reach before anything has authenticated
    /// it. An unhandled exception here fails the test, which is the outcome being ruled out.
    /// </summary>
    private static void DecodeAllReachable(byte[] payload)
    {
        PeerAppMessageCodec.LooksLike(payload);
        PeerAppMessageCodec.TryDecode(payload, out _);
        PeerPathMtu.LooksLike(payload);
        PeerPathMtu.Decode(payload);
    }

    private sealed class AdversarialVectors
    {
        public required string Name { get; init; }
        public required string Comment { get; init; }
        public required List<AdversarialCase> Cases { get; init; }
    }

    private sealed class AdversarialCase
    {
        public required string Name { get; init; }
        public required string Kind { get; init; }
        public required string PayloadHex { get; init; }
        public required string Expect { get; init; }
        public required string Why { get; init; }
    }
}

File: implementations/java/client/src/test/java/com/theshuai/specusclient/peer/SharedAdversarialVectorTests.java
package com.theshuai.specusclient.peer;

import com.fasterxml.jackson.databind.JsonNode;
import com.theshuai.common.util.JsonUtil;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HexFormat;

import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * The shared adversarial corpus in {@code protocol/test-vectors/adversarial-inputs.json}.
 *
 * <p>One implementation rejecting a hostile input does not mean the input is handled: an attacker
 * picks whichever node runs the most permissive implementation. Java, Go, .NET and Android all run
 * this same file, so a gap in one of them shows up as a failure rather than as a difference nobody
 * thought to look for.
 *
 * <p>The requirement per case is either a clean rejection or a value the caller can safely use —
 * never an unhandled exception, and never so slow that a hostile packet could stall a receive loop.
 */
class SharedAdversarialVectorTests {
    @Test
    void everySharedCaseIsHandled() throws IOException {
        JsonNode document = readVector("adversarial-inputs.json");
        JsonNode cases = document.get("cases");

        assertTrue(cases != null && cases.size() > 0,
                "the shared corpus is empty; the file was not read");

        for (JsonNode testCase : cases) {
            String name = testCase.get("name").asText();
            byte[] payload = HexFormat.of().parseHex(testCase.get("payloadHex").asText());

            long startedAt = System.nanoTime();
            decodeAllReachable(payload);
            long elapsedMillis = (System.nanoTime() - startedAt) / 1_000_000L;

            assertTrue(elapsedMillis < 1_000L, name + " took " + elapsedMillis
                    + "ms to decide; a hostile input must not stall a receive loop");
        }
    }

    /**
     * Feeds one payload to every decoder a datagram can reach before anything has authenticated
     * it. An unhandled exception here fails the test, which is the outcome being ruled out.
     */
    private static void decodeAllReachable(byte[] payload) {
        PeerAppMessageCodec.looksLike(payload);
        PeerAppMessageCodec.decode(payload);
        PeerPathMtu.looksLike(payload);
        PeerPathMtu.decode(payload);
    }

    private static JsonNode readVector(String name) throws IOException {
        Path current = Path.of("").toAbsolutePath();
        for (int depth = 0; current != null && depth < 8; depth++, current = current.getParent()) {
            Path candidate = current.resolve("protocol/test-vectors").resolve(name);
            if (Files.isRegularFile(candidate)) {
                return JsonUtil.readString(Files.readString(candidate));
            }
        }
        throw new IllegalStateException("cannot locate protocol vector: " + name);
    }
}
